linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala

```diff
@@ -201,19 +164,44 @@ private object MaxModuleAnalyzer {
       }
     }
 
-    def tags(): (Set[ModuleID], Set[ClassName]) = {
-      /* Remove dynamic paths to class that are also reached by a public module.
-       * However, only do this if there are other tags as well. Otherwise, this
-       * class will end up in a public module, but the dynamically loaded module
-       * will try to import it (but importing public modules is forbidden).
-       */
-      if (direct.size > 1 || direct != dynamic.keySet) {
-        direct.foreach(dynamic.remove(_))
-      }
+    def moduleID(internalModuleIDPrefix: String): ModuleID = {
+      if (direct.size == 1 && dynamic.isEmpty) {
+        /* Class is only used by a single public module. Put it there.
+         *
+         * Note that we must not do this if there are any dynamic modules
+         * requiring this class. Otherwise, the dynamically loaded module
+         * will try to import the public module (but importing public modules is
+         * forbidden).
+         */
+        direct.head
+      } else {
+        /* Class is used by multiple public modules and/or dynamic edges.
+         * Create a module ID grouping it with other classes that have the same
+         * dependees.
+         */
+        val digestBuilder = new SHA1.DigestBuilder
+
+        for (id <- direct.toList.sortBy(_.id))
+          digestBuilder.update(id.id.getBytes(StandardCharsets.UTF_8))
+
+        // Collect dynamic modules depending on this class.
+        val ends = {
+          val builder = immutable.SortedSet.newBuilder[ClassName]
+          // When iterating: Ignore path prefixes stored in `direct`.
+          for ((h, t) <- dynamic if !direct.contains(h))
```

Comment:
Note to self: I think this control flow changes grouping behavior. I should investigate.